[compiler-rt] Fix frame numbering for unparsable frames.

jschwartzentruber opened the pull request:
This can happen when JIT code is run, and we can't symbolize those frames, but they should remain numbered in the stack. An example spidermonkey trace:

```
    #0 0x564ac90fb80f  (/builds/worker/dist/bin/js+0x240e80f) (BuildId: 5d053c76aad4cfbd08259f8832e7ac78bbeeab58)
    #1 0x564ac9223a64  (/builds/worker/dist/bin/js+0x2536a64) (BuildId: 5d053c76aad4cfbd08259f8832e7ac78bbeeab58)
    #2 0x564ac922316f  (/builds/worker/dist/bin/js+0x253616f) (BuildId: 5d053c76aad4cfbd08259f8832e7ac78bbeeab58)
    #3 0x564ac9eac032  (/builds/worker/dist/bin/js+0x31bf032) (BuildId: 5d053c76aad4cfbd08259f8832e7ac78bbeeab58)
    #4 0x0dec477ca22e  (<unknown module>)
```

Without this change, the following symbolization is output:

```
    #0 0x55a6d72f980f in MOZ_CrashSequence /builds/worker/workspace/obj-build/dist/include/mozilla/Assertions.h:248:3
    #1 0x55a6d72f980f in Crash(JSContext*, unsigned int, JS::Value*) /builds/worker/checkouts/gecko/js/src/shell/js.cpp:4223:5
    #2 0x55a6d7421a64 in CallJSNative(JSContext*, bool (*)(JSContext*, unsigned int, JS::Value*), js::CallReason, JS::CallArgs const&) /builds/worker/checkouts/gecko/js/src/vm/Interpreter.cpp:501:13
    #3 0x55a6d742116f in js::InternalCallOrConstruct(JSContext*, JS::CallArgs const&, js::MaybeConstruct, js::CallReason) /builds/worker/checkouts/gecko/js/src/vm/Interpreter.cpp:597:12
    #4 0x55a6d80aa032 in js::jit::DoCallFallback(JSContext*, js::jit::BaselineFrame*, js::jit::ICFallbackStub*, unsigned int, JS::Value*, JS::MutableHandle<JS::Value>) /builds/worker/checkouts/gecko/js/src/jit/BaselineIC.cpp:1705:10
    #4 0x2c803bd8f22e  (<unknown module>)
```

The last frame has a duplicate number. With this change the numbering is correct:

```
    #0 0x5620c58ec80f in MOZ_CrashSequence /builds/worker/workspace/obj-build/dist/include/mozilla/Assertions.h:248:3
    #1 0x5620c58ec80f in Crash(JSContext*, unsigned int, JS::Value*) /builds/worker/checkouts/gecko/js/src/shell/js.cpp:4223:5
    #2 0x5620c5a14a64 in CallJSNative(JSContext*, bool (*)(JSContext*, unsigned int, JS::Value*), js::CallReason, JS::CallArgs const&) /builds/worker/checkouts/gecko/js/src/vm/Interpreter.cpp:501:13
    #3 0x5620c5a1416f in js::InternalCallOrConstruct(JSContext*, JS::CallArgs const&, js::MaybeConstruct, js::CallReason) /builds/worker/checkouts/gecko/js/src/vm/Interpreter.cpp:597:12
    #4 0x5620c669d032 in js::jit::DoCallFallback(JSContext*, js::jit::BaselineFrame*, js::jit::ICFallbackStub*, unsigned int, JS::Value*, JS::MutableHandle<JS::Value>) /builds/worker/checkouts/gecko/js/src/jit/BaselineIC.cpp:1705:10
    #5 0x349f24c7022e  (<unknown module>)
```

davidmrdavid commented on compiler-rt/lib/asan/scripts/asan_symbolize.py:
```diff
@@ -779,9 +788,13 @@ def __str__(self):
-            module_path=self.module_path
-            if self.module_path == self.module_path_for_symbolization
-            else "{} ({})".format(self.module_path_for_symbolization, self.module_path),
+            module_path=(
+                self.module_path
+                if self.module_path == self.module_path_for_symbolization
+                else "{} ({})".format(
+                    self.module_path_for_symbolization, self.module_path
+                )
+            ),
```
just checking - there's no change here, just formatting, right?
Was this intended or applied by some kind of auto-formatter? Just wondering if we should minimize the diff by reverting this.

As an aside - this assignment to `module_path` is moderately complex. I wonder if we could make it simpler by expanding the ternary expression into a proper `if-else` block. But that's probably out of scope.

jschwartzentruber replied:
This was the result of running `black` on the file, since I'm touching the file. I can revert if minimal diff is preferred.

davidmrdavid replied:
I can't say authoritatively one way or another, but I'd imagine a minimal diff will make it easier to approve.

davidmrdavid commented on compiler-rt/lib/asan/scripts/asan_symbolize.py:
```diff
@@ -38,6 +37,7 @@
 force_system_symbolizer = False
 
+
 # FIXME: merge the code that calls fix_filename().
```
is this `FIXME` still accurate? Otherwise, I'd expect this change to have no effect as it is no called as per the comment.

jschwartzentruber replied:
Ok, sorry. I don't know anything about the purpose of the function or the `FIXME`.

At the time the `FIXME` was added, the file already contained calls to `fix_filename` from the `symbolize` methods of a few `Symbolizer` subclasses. My guess is that the author hoped these callers could be merged into `super().symbolize()` instead of using a helper function. The implementations have gotten more complex since then, so I doubt the comment is still relevant.

davidmrdavid commented on compiler-rt/lib/asan/scripts/asan_symbolize.py:
```diff
@@ -507,20 +507,29 @@ def symbolize_address(self, addr, binary, offset, arch):
                 self.frame_no += 1
-            return result
+                return [rewritten_line]
+            # Not a frame line so don't increment the frame counter.
```
just confirming - if it's not a frame line, then what could it be? I'm wondering if that should be made explicit in the comments.

jschwartzentruber replied:
It could be anything else present in the log -- which could either be non-frame lines in the sanitizer log or other process output if the log is from stderr.

davidmrdavid replied:
Noted. I also see this comment exists as-is in the previous diff, so I think it's fine to leave as-is. Resolving.

davidmrdavid commented on compiler-rt/lib/asan/scripts/asan_symbolize.py:
```diff
@@ -507,20 +507,29 @@ def symbolize_address(self, addr, binary, offset, arch):
+            # replace the frame counter so the stack is still consistent.
+            unknown_stack_frame_format = r"^( *#([0-9]+) +)(0x[0-9a-f]+) +.*"
+            match = re.match(unknown_stack_frame_format, self.current_line)
+            if match:
+                rewritten_line = (
+                    self.current_line[: match.start(2)]
+                    + str(self.frame_no)
+                    + self.current_line[match.end(2) :]
```
dismissible nit, I would recommend introducing a variable for `match.start(2)`, to help with readability. The expression is already a bit complex, _in my opinion_.

```python
           if match:
                # update the frame counter
                frame_no_index = match.start(2)
                rewritten_line= (
                    self.current_line[: frame_no_index ]
                    + str(self.frame_no)
                    + self.current_line[frame_no_index  :]
```

jschwartzentruber replied:
It would be better to use a named match group, so the calls would be `match.start("frame_no")` and `match.end("frame_no")`. This clutters the regex though, which is usually the harder bit to understand.

davidmrdavid commented on compiler-rt/lib/asan/scripts/asan_symbolize.py:
```diff
@@ -507,20 +507,29 @@ def symbolize_address(self, addr, binary, offset, arch):
-            for symbolized_frame in symbolized_lines:
-                result.append(
-                    "    #%s %s" % (str(self.frame_no), symbolized_frame.rstrip())
+            # If it is an unparsable frame, but contains a frame counter and address
+            # replace the frame counter so the stack is still consistent.
```
question: something here is a bit strange - it seems to me that we're fixing the frame counter _after_ the frame string has been generated, so it feels like there would be an earlier point in the process where the frame counter is not being incremented. Do we know where that is, and why we can't update it _there_?

davidmrdavid replied:
GH is acting up so I seem to be unable to resolve this comment :/

davidmrdavid reviewed (commented):
Just an OSS contributor, I don't have commit access, but wanted to leave some suggestions in case they help. Hope they're helpful!

davidmrdavid commented:
Thanks @thurstond, that would be wonderful! I'll look to follow those instructions later :) Thanks!

davidmrdavid commented:
@jschwartzentruber - is there a simple test (either manual or automated) that we can run to validate these changes?

thurstond commented:
> **Update:** @thurstond - I created an issue here, thanks! #151001

I've commented my approval there, and also pinged some reviewers to request the second approval :-)

jschwartzentruber commented:
> @jschwartzentruber - is there a simple test (either manual or automated) that we can run/add to the PR to validate these changes?

I looked for existing automated tests but didn't see any. Happy to add a test if there is a framework for it.

The manual steps I used to reproduce the issue use spidermonkey crashing in JIT code as an example:

```
pipx run fuzzfetch --asan --target js -n js-asan
echo "while (! inJit()) {}; crash()" | ASAN_OPTIONS=symbolize=0 ./js-asan/dist/bin/js 2>&1 | python ~/src/llvm-project/compiler-rt/lib/asan/scripts/asan_symbolize.py -d
```

thurstond commented:
> I looked for existing automated tests but didn't see any. Happy to add a test if there is a framework for it.

Maybe something similar to `compiler-rt/test/asan/TestCases/Posix/asan_symbolize_script/plugin_wrong_frame_number_bug.cpp`?

thurstond reviewed (approved):
Ideally, please add a test case (bonus points if the test case is precommitted in a separate PR, and this PR rebased on top of that), but I'd be happy for this to be merged as-is.

Thanks for the patch!

jschwartzentruber committed:
[compiler-rt] Fix frame numbering for unparsable frames.

This can happen when JIT code is run, and we can't symbolize those
frames, but they should remain numbered in the stack.
